Bugfix for windowFunnel's "strict" mode.

achimbab opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Bugfix for windowFunnel's "strict" mode. This fixes #27469.

otheng03 committed:
Bugfix for windowFunnel's "strict" mode.

BiteTheDDDDt commented:
Is it better to change several `return` conditions in the loop to `break`, so that the mixing of multiple `Mode` will not easily affect each other?
`Mode` only affects the timing of breaks, and their definitions will be more precise.
They will have uniform performance, and the result will be calculated at the end of the function.

achimbab commented:
@BiteTheDDDDt 
That great, but when it comes to mix of two options like below, splitting each modes can be lead to many combinations of options.

```SQL
select 
  user, 
  windowFunnel(86400, 'strict', 'strict_order')(dt, event='a', event='b', event='c') as s 
from funnel_test_strict_order 
group by user 
order by user 
format JSONCompactEachRow;
```

mathalex reviewed (commented):
The code is OK. Few comments:
1. Line 153: the algorithm itself of course works in O(n) time. However, the overall function, before which you posted a comment, works in O(n * log(n)) time due to the sort function. Probably, it would be better to mention it.
2. (FYI, mostly for the documentation team) Documentation says that `strict` parameter just skip non-unique events, but, as I can see in the code, repeating event interrupts further processing.
3. Maybe if you have added the `strict_deduplication` parameter it would be better to delete the `strict` parameter at all since one can think that this option just turns on all three strict options (so you can not delete it but make this functionality).

otheng03 committed:
Fix docs and comments for windowFunnel and remove strict option

otheng03 committed:
Merge branch 'master' into bugfix_wf_strict

achimbab commented:
@mathalex 
> Line 153: the algorithm itself of course works in O(n) time. However, the overall function, before which you posted a comment, works in O(n * log(n)) time due to the sort function. Probably, it would be better to mention it.

I fixed the comment.

> (FYI, mostly for the documentation team) Documentation says that strict parameter just skip non-unique events, but, as I can see in the code, repeating event interrupts further processing.

I fixed the document.

> Maybe if you have added the strict_deduplication parameter it would be better to delete the strict parameter at all since one can think that this option just turns on all three strict options (so you can not delete it but make this functionality).

I removed the `strict` option and fixed code to throw an exception in order to notify users that the `strict` was replaced with `strict_deduplication`.

mathalex commented on src/AggregateFunctions/AggregateFunctionWindowFunnel.h:
```diff
@@ -150,7 +150,7 @@ class AggregateFunctionWindowFunnel final
     /// The level path must be 1---2---3---...---check_events_size, find the max event level that satisfied the path in the sliding window.
     /// If found, returns the max event level, else return 0.
-    /// The Algorithm complexity is O(n).
+    /// The algorithm works in O(n * log(n)) time.
```
OK, but maybe it would be better to mention that this asymptotic is only due to sorting. The described in the comment algorithm is works blatantly in O(n), so the reader can be confused. Something like this:
```suggestion
    /// The algorithm works in O(n) time, but the overall function works in O(n * log(n)) due to sorting.
```

mathalex commented on src/AggregateFunctions/AggregateFunctionWindowFunnel.h:
```diff
@@ -226,18 +226,20 @@ class AggregateFunctionWindowFunnel final
             else if (option == "strict_increase")
                 strict_increase = true;
+            else if (option == "strict")
+                throw Exception{"strict is replaced with strict_deduplication in Aggregate function " + getName(), ErrorCodes::BAD_ARGUMENTS};
```
This is a good solution until documentation in other languages will not be fixed.

Commit:
Update docs/en/sql-reference/aggregate-functions/parametric-functions.md

Co-authored-by: Alexey Boykov <33257111+mathalex@users.noreply.github.com>

Commit:
Update src/AggregateFunctions/AggregateFunctionWindowFunnel.h

Co-authored-by: Alexey Boykov <33257111+mathalex@users.noreply.github.com>
